File: web/src/fill-engine/bootstrap.ts
export interface ConfidenceInterval {
  median: number
  lower: number
  upper: number
}

export interface BootstrapResult {
  expectancy: ConfidenceInterval
  maxDrawdown: ConfidenceInterval
  sharpe: ConfidenceInterval
  iterations: number
  confidence: number
  seed: number
}

export interface BootstrapOptions {
  iterations?: number
  confidence?: number
  seed?: number
}

function seededRandom(seed: number): () => number {
  let state = seed >>> 0
  return () => {
    state = (state * 1_664_525 + 1_013_904_223) >>> 0
    return state / 4_294_967_296
  }
}

function average(values: number[]): number {
  return values.length === 0 ? 0 : values.reduce((sum, value) => sum + value, 0) / values.length
}

function maxDrawdown(values: number[]): number {
  let equity = 0
  let peak = 0
  let maximum = 0
  for (const value of values) {
    equity += value
    peak = Math.max(peak, equity)
    maximum = Math.min(maximum, equity - peak)
  }
  return maximum
}

function sharpe(values: number[]): number {
  if (values.length < 2) return 0
  const mean = average(values)
  const variance = values.reduce((sum, value) => sum + (value - mean) ** 2, 0) / (values.length - 1)
  const deviation = Math.sqrt(Math.max(0, variance))
  return deviation === 0 ? 0 : mean / deviation * Math.sqrt(values.length)
}

function quantile(sorted: number[], percentile: number): number {
  if (sorted.length === 0) return 0
  const position = Math.min(sorted.length - 1, Math.max(0, percentile * (sorted.length - 1)))
  const lower = Math.floor(position)
  const upper = Math.ceil(position)
  const weight = position - lower
  return (sorted[lower] ?? 0) * (1 - weight) + (sorted[upper] ?? 0) * weight
}

function interval(values: number[], confidence: number): ConfidenceInterval {
  const sorted = values.toSorted((a, b) => a - b)
  const tail = (1 - confidence) / 2
  return { lower: quantile(sorted, tail), median: quantile(sorted, 0.5), upper: quantile(sorted, 1 - tail) }
}

export function bootstrapAnalytics(values: number[], options: BootstrapOptions = {}): BootstrapResult {
  const r = values.filter(Number.isFinite)
  const iterations = Math.min(50_000, Math.max(100, Math.round(options.iterations ?? 10_000)))
  const confidence = Math.min(0.99, Math.max(0.5, options.confidence ?? 0.9))
  const seed = options.seed ?? 20_250_816
  if (r.length === 0) {
    const empty = { median: 0, lower: 0, upper: 0 }
    return { expectancy: empty, maxDrawdown: empty, sharpe: empty, iterations, confidence, seed }
  }
  const random = seededRandom(seed)
  const expectancy: number[] = []
  const drawdown: number[] = []
  const sharpeValues: number[] = []
  const sample = new Array<number>(r.length)
  for (let iteration = 0; iteration < iterations; iteration += 1) {
    for (let index = 0; index < r.length; index += 1) sample[index] = r[Math.floor(random() * r.length)] ?? 0
    expectancy.push(average(sample))
    drawdown.push(maxDrawdown(sample))
    sharpeValues.push(sharpe(sample))
  }
  return {
    expectancy: interval(expectancy, confidence),
    maxDrawdown: interval(drawdown, confidence),
    sharpe: interval(sharpeValues, confidence),
    iterations,
    confidence,
    seed,
  }
}

File: internal/analytics/bootstrap.go
package analytics

import (
	"context"
	"math"
)

// bootstrapCancellationCheckInterval is how often (in iterations) a long
// bootstrap run checks ctx for cancellation — frequent enough that a
// canceled request stops promptly, infrequent enough that the check
// itself is not the bottleneck at 50,000 iterations.
const bootstrapCancellationCheckInterval = 100

// ConfidenceInterval is a [Lower,Upper] interval around Median.
type ConfidenceInterval struct {
	Median float64 `json:"median"`
	Lower  float64 `json:"lower"`
	Upper  float64 `json:"upper"`
}

// BootstrapResult is BootstrapAnalytics' result. ExpectancyR/MaxDrawdownR
// are both R-multiple scales (not percent); MaxDrawdownR is peak-to-trough
// on the cumulative-R curve, so it is <= 0.
type BootstrapResult struct {
	ExpectancyR  ConfidenceInterval `json:"expectancyR"`
	MaxDrawdownR ConfidenceInterval `json:"maxDrawdownR"`
	Sharpe       ConfidenceInterval `json:"sharpe"`
	Iterations   int                `json:"iterations"`
	Confidence   float64            `json:"confidence"`
	Seed         int64              `json:"seed"`
}

// BootstrapOptions mirrors bootstrap.ts' BootstrapOptions, except Seed is
// always a concrete value here rather than optional: per policy P8, "no
// seed supplied" is resolved to a freshly generated one by the httpapi
// layer (same randomSeed() helper the Monte Carlo simulator uses) before
// this pure function ever runs — this package always knows, and always
// reports, the seed actually used.
type BootstrapOptions struct {
	Iterations *int
	Confidence *float64
	Seed       int64
}

const (
	bootstrapDefaultIterations = 10_000
	bootstrapMinIterations     = 100
	bootstrapMaxIterations     = 50_000
	bootstrapDefaultConfidence = 0.9
	bootstrapMinConfidence     = 0.5
	bootstrapMaxConfidence     = 0.99
)

func sharpeOf(values []float64) float64 {
	if len(values) < 2 {
		return 0
	}
	deviation := sampleStdDev(values)
	if deviation == 0 {
		return 0
	}
	return average(values) / deviation * math.Sqrt(float64(len(values)))
}

// BootstrapAnalytics resamples values (with replacement) Iterations times,
// computing expectancy/maxDrawdown/sharpe on each resample, then reports
// the Confidence-level interval for each. Deterministic: the same
// values+Seed always produce byte-identical output. ctx is checked for
// cancellation every bootstrapCancellationCheckInterval iterations — a
// canceled request returns ctx.Err() promptly instead of finishing all
// 50,000 possible iterations first.
func BootstrapAnalytics(ctx context.Context, values []float64, opts BootstrapOptions) (BootstrapResult, error) {
	iterations := bootstrapDefaultIterations
	if opts.Iterations != nil {
		iterations = *opts.Iterations
	}
	if iterations < bootstrapMinIterations {
		iterations = bootstrapMinIterations
	}
	if iterations > bootstrapMaxIterations {
		iterations = bootstrapMaxIterations
	}

	confidence := bootstrapDefaultConfidence
	if opts.Confidence != nil {
		confidence = *opts.Confidence
	}
	if confidence < bootstrapMinConfidence {
		confidence = bootstrapMinConfidence
	}
	if confidence > bootstrapMaxConfidence {
		confidence = bootstrapMaxConfidence
	}

	r := finiteFloats(values)
	if len(r) == 0 {
		empty := ConfidenceInterval{}
		return BootstrapResult{
			ExpectancyR: empty, MaxDrawdownR: empty, Sharpe: empty,
			Iterations: iterations, Confidence: confidence, Seed: opts.Seed,
		}, nil
	}

	rng := newSeededRandom(opts.Seed)
	expectancy := make([]float64, iterations)
	drawdown := make([]float64, iterations)
	sharpeValues := make([]float64, iterations)
	sample := make([]float64, len(r))
	for iteration := 0; iteration < iterations; iteration++ {
		if iteration%bootstrapCancellationCheckInterval == 0 {
			if err := ctx.Err(); err != nil {
				return BootstrapResult{}, err
			}
		}
		for i := range sample {
			idx := int(rng.next() * float64(len(r)))
			if idx >= len(r) {
				idx = len(r) - 1
			}
			sample[i] = r[idx]
		}
		expectancy[iteration] = average(sample)
		drawdown[iteration] = maxDrawdownOfSeries(sample)
		sharpeValues[iteration] = sharpeOf(sample)
	}

	return BootstrapResult{
		ExpectancyR:  confidenceIntervalOf(expectancy, confidence),
		MaxDrawdownR: confidenceIntervalOf(drawdown, confidence),
		Sharpe:       confidenceIntervalOf(sharpeValues, confidence),
		Iterations:   iterations,
		Confidence:   confidence,
		Seed:         opts.Seed,
	}, nil
}
